src/backend/buffer.rs

```diff
@@ -123,30 +123,39 @@ where
         unsafe { str::from_utf8_unchecked(str_bytes) }
     }
 
-    /// Pushes the given value onto the buffer with `var7` encoding.
+    /// Pushes the given length value onto the buffer with `var7` encoding.
+    /// Ensures there's enough capacity for the `var7` encoded length and
+    /// the following string bytes ahead of pushing.
     ///
     /// Returns the amount of `var7` encoded bytes.
     #[inline]
-    fn encode_var_usize(&mut self, value: usize) -> usize {
-        encode_var_usize(&mut self.buffer, value)
+    fn try_encode_var_length(&mut self, length: usize) -> Result<()> {
+        let add_len = length + calculate_var7_size(length);
+        self.buffer.try_reserve(add_len)?;
+        encode_var_usize(&mut self.buffer, length);
+        Ok(())
     }
 
-    /// Pushes the given string into the buffer and returns its span.
+    /// Pushes the given string into the buffer and returns its span on success.
     ///
-    /// # Panics
+    /// # Errors
     ///
-    /// If the backend ran out of symbols.
-    fn push_string(&mut self, string: &str) -> S {
-        let symbol = self.next_symbol();
+    /// Returns an [`Error`] if the backend ran out of symbols or memory.
+    fn try_push_string(&mut self, string: &str) -> Result<S> {
+        let symbol = self.try_next_symbol()?;
         let str_len = string.len();
         let str_bytes = string.as_bytes();
-        self.encode_var_usize(str_len);
-        self.buffer.extend(str_bytes);
+        self.try_encode_var_length(str_len)?;
+        // try_encode_var_length ensures there's enough space left for str_bytes
+        self.buffer.extend_from_slice(str_bytes);
         self.len_strings += 1;
-        symbol
+        Ok(symbol)
     }
 }
 
+/// According to google the approx. word length is 5.
+const DEFAULT_STR_LEN: usize = 5;
```

Comment:
We should really get rid of this thing entirely. My bad to have introduced them in the first place. 🙈

Reply:
By that, do you mean to remove `with_capacity` from the Backend altogether?